plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/impl/BaseFlightProducer.java

```diff
@@ -0,0 +1,283 @@
+            );
+            return null;
+        }
+    }
+
+    private void processStream(
+        FlightStreamManager.StreamProducerHolder streamProducerHolder,
+        StreamProducer.BatchedJob<VectorSchemaRoot> batchedJob,
+        CallContext context,
+        ServerStreamListener listener
+    ) {
+        listener.setOnCancelHandler(batchedJob::onCancel);
+        BackpressureStrategy backpressureStrategy = new BaseBackpressureStrategy(null, batchedJob::onCancel);
+        backpressureStrategy.register(listener);
+
+        StreamProducer.FlushSignal flushSignal = createFlushSignal(batchedJob, listener, backpressureStrategy);
+
+        try (VectorSchemaRoot root = streamProducerHolder.getRoot()) {
+            listener.start(root);
+            batchedJob.run(root, flushSignal);
+            listener.completed();
+        }
+    }
+
+    private StreamProducer.FlushSignal createFlushSignal(
+        StreamProducer.BatchedJob<VectorSchemaRoot> batchedJob,
+        ServerStreamListener listener,
+        BackpressureStrategy backpressureStrategy
+    ) {
+        return (timeout) -> {
+            BackpressureStrategy.WaitResult result = backpressureStrategy.waitForListener(timeout.millis());
+            switch (result) {
+                case READY:
+                    listener.putNext();
+                    break;
+                case TIMEOUT:
+                    batchedJob.onCancel();
+                    listener.error(CallStatus.TIMED_OUT.withDescription("Stream deadline exceeded").toRuntimeException());
+                    break;
+                case CANCELLED:
+                    batchedJob.onCancel();
+                    listener.error(CallStatus.CANCELLED.withDescription("Stream cancelled by client").toRuntimeException());
+                    break;
+                default:
+                    batchedJob.onCancel();
+                    logger.error("Unexpected backpressure result: {}", result);
+                    listener.error(CallStatus.INTERNAL.withDescription("Error waiting for client: " + result).toRuntimeException());
+            }
+        };
+    }
+
+    private void handleCancellation(StreamProducer.BatchedJob<VectorSchemaRoot> batchedJob, ServerStreamListener listener) {
+        try {
+            batchedJob.onCancel();
+            listener.error(CallStatus.CANCELLED.withDescription("Stream cancelled before processing").toRuntimeException());
+        } catch (Exception e) {
+            logger.error("Error during cancellation handling", e);
+            listener.error(
+                CallStatus.INTERNAL.withDescription("Error during cancellation: " + e.getMessage()).withCause(e).toRuntimeException()
+            );
+        }
+    }
+
+    private FlightInfo getLocalFlightInfo(StreamTicket streamTicket, FlightDescriptor descriptor) {
+        Optional<FlightStreamManager.StreamProducerHolder> streamProducerHolder = streamManager.getStreamProducer(streamTicket);
```

Comment:
```suggestion
        logger.debug("Failed to determine location for node: {}", streamTicket.getNodeId());
                throw CallStatus.UNAVAILABLE.withDescription("Internal error determining location").toRuntimeException();
            }
```

Reply:
nice! I need to work on my "Optional" skills :D